### ISSUE TITLE
FFI Thread-Safe Callbacks

### ISSUE THREAD

**Issue by DjDeveloperr** [denoland/deno#13573]
I propose adding a new API `Deno.ThreadSafeCallback` to create C-style callbacks which can be passed down to FFI.

### Why

Currently, #13162 adds support for passing JS callbacks wrapped in a C function to FFI. However, that will not work for callbacks that will be moved to another thread as V8 Isolate cannot be accessed from other threads. Instead, we could implement a new mechanism for creating C callbacks which do not use any V8 API, as a result which can be called from any thread but handled again on Isolate thread. It is also much easier & safer to implement than trying to do same with JS callbacks using V8 API.

### Proposal

```ts
declare class ThreadSafeCallback<Fn extends Deno.ForeignFunction> {
  readonly pointer: Deno.UnsafePointer;
  
  // It will create a new callback resource that is basically just a Rust channel
  // and a closure. The closure will create a new Rust channel that will wait for return
  // value and sends the arguments & new channel's sender end to the previous
  // Rust channel. Further logic is handled by `next`.
  constructor(def: Fn);

  // This will spawn a tokio blocking task to wait for the function to be called
  // on other thread (as such isolate thread is not blocked). Basically, it will poll
  // for new messages in the Rust channel.
  async next(): Promise<ThreadSafeCallbackContext<Fn>>;

  // This is just to unref any current or future async op call made by
  // `next` function because it would by default cause Deno event loop
  // to not exit until the function has not been called.
  unref(): void;

  // Since it is a resource, we'll explicitly manage it's lifetime.
  close(): void;

  // For allowing `for await of` loops to handle function calls (basically wraps the `next` function)
  *[Symbol.asyncIterator](): AsyncIterableIterator<ThreadSafeCallbackContext<Fn>>;
}

declare class ThreadSafeCallbackContext<Fn extends Deno.ForeignFunction> {
  readonly arguments: MapFunctionToArguments<Fn>;

  // This will send return value to the another Rust channel created
  // when the callback was called, causing the wait for it to end.
  return(value: MapNativeTypeToJS<Fn["result"]>): void;
}
```

### Example usage

```ts
// Creating:
const callback = new Deno.ThreadSafeCallback({
  parameters: ["i32", "i32"],
  result: "i32",
});

// Pass callback.pointer to FFI

// Handling:
callback.next().then((ctx) => {
  ctx.return(ctx.arguments[0] + ctx.arguments[1]);
});
// or like this:
for await (const ctx of callback) {
  ctx.return(ctx.arguments[0] + ctx.arguments[1]);
}
```

### But but...

This approach has a problem. The created callback function cannot be called from main thread (or thread same as the Isolate), otherwise it deadlocks waiting for return value on same thread. One solution is to have an hybrid approach which calls either directly V8 function or falls back to this when called from other thread.

**Comment by aapoalas** [comment]
I very much like this idea. I also toyed around with the idea of trying to do this with a reader pattern but that probably doesn't really make much sense.

One option for the `return` and might be to include the callback in the constructor instead of using the explicit `ThreadSafeCallbackContext` class to "simulate" the callback. If one thinks of callbacks in FFI, there is no way to "cram in" multiple callbacks into that one callback parameter. The API proposed here makes it possible for one FFI call with one callback parameter to split into multiple handlers, calling multiple returns which would definitely be an error of some sort.

A point of optimization is that a lot of FFI callbacks will likely not have a return value. For these, I would expect Deno to internally optimize to not wait for any async callback finishing.

**Comment by aapoalas** [comment]
One further thought I had: At present I believe that #13162 callbacks do not survive beyond the actual FFI call, ie. even if an FFI call takes a callback that will be called from the same thread, it cannot keep that callback and call it later.

This seems like an understandable limitation based on my limited understanding of the Rust code underneath and how it needs to manage the lifetime of the callback info etc. However, a lot of FFI libraries will not function with this limitation.

For these cases in order to make the lifetimes explicitly handled from JS-side, an explicit callback class like `ThreadSafeCallback` might be needed.

**Comment by DjDeveloperr** [comment]
> One option for the `return` and might be to include the callback in the constructor instead of using the explicit `ThreadSafeCallbackContext` class to "simulate" the callback.

Sounds good to me, it would make more sense to just do that than a explicit call context

> A point of optimization is that a lot of FFI callbacks will likely not have a return value. For these, I would expect Deno to internally optimize to not wait for any async callback finishing.

I agree in most use cases callbacks don't even need to return something. In case we don't have a return value, there wouldn't even be a deadlock (which is caused by waiting for return value on same thread) as I mentioned in issue. But that *might* cause issues because a library may expect callback to be executed synchronously, which would not really be if we don't wait for the function to return.

> This seems like an understandable limitation based on my limited understanding of the Rust code underneath and how it needs to manage the lifetime of the callback info etc.

I don't think there's any limitation, we can just have a callback resource for that using Deno's resource table.

I actually have a POC for this API built using FFI, only thing it is missing is `unref`ing the async op is not possible, so a `next` call will prevent Deno from exiting.. which can likely be fixed if its implemented into Deno itself since I think there is an internal API to unref async ops.

**Comment by aapoalas** [comment]
> I agree in most use cases callbacks don't even need to return something. In case we don't have a return value, there wouldn't even be a deadlock (which is caused by waiting for return value on same thread) as I mentioned in issue. But that _might_ cause issues because a library may expect callback to be executed synchronously, which would not really be if we don't wait for the function to return.

Yeah, what I meant by this was that if a `ThreadSafeCallback` was provided an `async` function for the callback while the definition of the callback expected no return value, then Deno would just call the JS callback and disregard the Promise returned, yielding to the caller immediately after without waiting for the Promise to finish first.

That is, if `async` functions as callbacks are allowed.

> I don't think there's any limitation, we can just have a callback resource for that using Deno's resource table.
> 
> I actually have a POC for this API built using FFI, only thing it is missing is `unref`ing the async op is not possible, so a `next` call will prevent Deno from exiting.. which can likely be fixed if its implemented into Deno itself since I think there is an internal API to unref async ops.

Do you have the code for this available somewhere?

**Comment by DjDeveloperr** [comment]
> Do you have the code for this available somewhere?

Just made a repo: https://github.com/DjDeveloperr/deno_threadsafe_callback_poc

**Comment by aapoalas** [comment]
Cool stuff! I wonder if the code as it currently stands wouldn't cause issues with my use-case, though.

In my case I have two callbacks, one of them needs to be thread-safe whereas the other is always called in response to another FFI call and from the same thread.

So my FFI calls would look something like this:
```
const notifyCallback = new ThreadSafeCallback(() => {
  // Notify callback gets called when new stuff is available in queue.
  // We can delay the call to debounce if we want to.
  ffi.drainQueue();
});

// Register the notify callback
ffi.setNotifyCallback(notifyCallback);

// Elsewhere: This callback will always be called from the main thread after calling `drainQueue`, possibly multiple times. It's saved on the foreign FFI side of things.
const streamCallback = new ThreadSafeCallback((status: number, data: Deno.UnsafePointer) => {
  if (status === 0) {
    // Everything is OK, data can be read
    readDataFromPointer(data);
  } else {
     // Something went wrong
    handleErrorFromPointer(data);
  }
});
ffi.startStream(streamOpts, streamCallback);
```

With the POC, I would have to call `streamCallback.then()` which would mean going to a different thread where the callback's parameters would be moved to. That is, if I'm understanding the code correctly. Threading is a bit unfamiliar territory for me. Anyway, I'm not sure if the library could actually deal with that.

**Comment by DjDeveloperr** [comment]
It can deal with that if you don't have any return values, and JS callback is easy to simulate as ThreadSafeCallback too, but I guess only problem you'd face is it will prevent Deno from exiting. I'll try implementing this in Deno itself sometime and see if I can fix the issue

**Comment by aapoalas** [comment]
Oh yeah, I got confused somehow. Yeah, this should work and I'll have to try it out.

Most, though not all, of my `ThreadSafeCallback` use-cases should actually prevent Deno from exiting. Only one or two I believe are such that I want to explicitly `unref` them.

**Comment by aapoalas** [comment]
I created a more ... formal is the wrong word but more wordy proposal with some code examples here: https://github.com/aapoalas/deno-stable-ffi

**Comment by littledivy** [comment]
@aapoalas' patch has landed #14942
